### ISSUE TITLE
Automatic Differentiation

### ISSUE THREAD

**Issue by ChufanSuki** [ChufanSuki/read-paper-and-code#142]
[Differentiation Notation](https://en.wikipedia.org/wiki/Notation_for_differentiation)

- Leibniz's notation: $\frac{d^nf}{dx^n}$, $df=f'(x) \cdot dx$
- Lagrange's notation: $f^{(n)}(x)$
- D-notation: $(Df)(x)=\frac{df}{dx}(x)$
- Newton's notation: $\dot{y}$

Confusing Terminology

- Automatic differentiation(autodiff)
- Backpropagation
- Autograd

Backpropagation is an instance of reverse mode automatic differentiation.

```math
\frac{\partial f(\mathbf{x})}{\partial x_i} \approx \frac{f\left(\mathbf{x}+h \mathbf{e}_i\right)-f(\mathbf{x})}{h}
```

Autodiff is not finite differences(only use it for testing). Numerical approximations of derivatives are inherently ill-conditioned and unstable("thou shalt not add small numbers to big numbers", and "thou shalt not subtract numbers which are approximately equal").  This is due to the introduction of truncation and round-off error.

Truncation error is the error of approximation, or inaccuracy, one gets from $h$ not actually being zero. It is proportional to a power of $h$.

Round-off error is the inaccuracy one gets from valuable low-order bits of the final answer having to compete for machine-word space with high-order bits of $f(\mathbf{x} + h \mathbf{e}_i)$ and $f(\mathbf{x})$, which the computer has to store just until they cancel in the subtraction at the end. Round-off error is inversely proportional to a power of $h$.

Autodiff is not symbolic differentiation. 

An autodiff system will convert the program into a sequence of _primitive operations(ops)_ which have specified routines for computing derivatives. In this representation, backprop can be done in a completely mechanical way. This is “just” a clever and efficient use of the Chain Rule for derivatives.

Example: Jax(autograd), Pytorch, MxNet

**Comment by ChufanSuki** [comment]
Building the Computation Graph

Node: A `Node` is an abstract class that represents an operation taking zero or more input `Variable`s and producing zero or more output `Variable`s.

Autograd's Node class (defined in [tracer.py](https://github.com/mattjj/autodidact/blob/c32947fbd7f130d791ab1e277a0feed05cb3da5f/autograd/tracer.py#L35C1-L60C20) ) represents a node of the computation graph. It has attributes:

- `value`, the actual value computed on a particular set of inputs
- `fun`, the primitive operation defining the node
- `args` and `kwargs`, the arguments the op was called with
- `parents`, the parent Nodes

**Comment by ChufanSuki** [comment]
![image](https://github.com/ChufanSuki/read-paper-and-code/assets/50797312/c4866474-c734-4f8b-9a6c-f71846d3fea3)

**Comment by ChufanSuki** [comment]
The backprop equation (single child node) can be written as a vector-Jacobian product (VJP):
```math
\overline{x_j}=\sum_i \overline{y_i} \frac{\partial y_i}{\partial x_j} \quad \overline{\mathbf{x}}=\overline{\mathbf{y}}^{\top} \mathbf{J}
```

That gives a row vector. We can treat it as a column vector by taking
```math
\overline{\mathbf{x}}=\mathbf{J}^{\top} \overline{\mathbf{y}}
```

we never explicitly construct the Jacobian. It’s usually simpler and more efficient to compute the VJP directly.

Consider a naïve backprop implementation where the $\mathbf{z}$ module needs to compute $\overline{\mathbf{z}}$ using the formula:

![image](https://github.com/ChufanSuki/read-paper-and-code/assets/50797312/1f7fd8f5-8b7f-49fe-98b1-544a18ebafaa)

```math
\overline{\mathbf{z}}=\frac{\partial \mathbf{r}}{\partial \mathbf{z}} \overline{\mathbf{r}}+\frac{\partial \mathbf{s}}{\partial \mathbf{z}} \overline{\mathbf{s}}+\frac{\partial \mathbf{t}}{\partial \mathbf{z}} \overline{\mathbf{t}}
```

This breaks modularity, since $\mathbf{z}$ needs to know how it's used in the network in order to compute partial derivatives of $\mathbf{r}, \mathbf{s}$, and $\mathbf{t}$.

Backprop as message passing:

![image](https://github.com/ChufanSuki/read-paper-and-code/assets/50797312/e7cd6536-4185-42ac-98e6-81d386334f65)

- Each node receives a bunch of messages from its children, which it aggregates to get its error signal. It then passes messages to its parents.
- Each of these messages is a VJP.
- This formulation provides modularity: each node needs to know how to compute its outgoing messages, i.e. the VJPs corresponding to each of its parents (arguments to the function).
- The implementation of z doesn’t need to know where z came from.

**Comment by ChufanSuki** [comment]
Read more:

1. https://github.com/renmengye/tensorflow-forward-ad/issues/2
2. https://github.com/HIPS/autograd/pull/55

**Comment by ChufanSuki** [comment]
### mxnet's AD

SimpleBind:

- `arg`: input arguments array
- `req`: requirement type of gradient saving
- `aux`: op's internal state array

**Comment by ChufanSuki** [comment]
Jax's AD

https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html

**Comment by ChufanSuki** [comment]
PyTorch' AD
